src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/MonadBlueprint.hs

```diff
@@ -104,19 +105,22 @@ class Monad m => MonadBlueprint i a m | m -> i, m -> a where
     newAssigned :: ClosedPoly i a -> m i
     newAssigned p = newConstrained (\x i -> p x - x i) p
 
-instance Arithmetic a => MonadBlueprint Integer a (State (ArithmeticCircuit a)) where
+instance Arithmetic a => MonadBlueprint Natural a (State (ArithmeticCircuit a)) where
     input = acOutput <$> I.input
 
     output i = gets (\r -> r { acOutput = i })
 
     runCircuit r = modify (<> r) $> acOutput r
 
     newConstrained
-        :: NewConstraint Integer a
-        -> Witness Integer a
-        -> State (ArithmeticCircuit a) Integer
+        :: NewConstraint Natural a
+        -> Witness Natural a
+        -> State (ArithmeticCircuit a) Natural
     newConstrained new witness = do
-        let s = sources witness `Set.difference` sources (`new` (-1))
+        let ws = sources witness
```

Comment:
@TurtlePU Can you elaborate here a little bit please?

Reply:
`s` is meant to be a set of variables used in a witness not present in a constraint. To compute the set of variables not present in a constraint, we first have to pass a "new variable" which we are generating, but here it is enough to pass a variable which definitely would not be present in a witness